### PR TITLE
feat: create conversation

### LINKED ISSUE #3752: port: add create conversation to CloudAdapter (#5562)
The changes in [add create conversation to CloudAdapter (#5562)](https://github.com/microsoft/botbuilder-dotnet/pull/5562) may need to be ported to maintain parity with `microsoft/botbuilder-dotnet`.

<blockquote>
Fixes #5488

Also updates the abstract class to avoid casts in the application code.
</blockquote>

Please review and, if necessary, port the changes.

### PR DESCRIPTION
**Pull request by joshgummersall**
Fixes #3752

### CODE REVIEW

File: libraries/botbuilder-core/src/cloudAdapterBase.ts [hunk 1]
```diff
@@ -176,6 +179,82 @@ export abstract class CloudAdapterBase extends BotAdapter {
...
+        botAppId: string,
+        channelId: string,
+        serviceUrl: string,
+        audience: string,
+        conversationParameters: ConversationParameters,
+        logic: (context: TurnContext) => Promise<void>
+    ): Promise<void> {
+        // Create a ClaimsIdentity, to create the connector and for adding to the turn context.
+        const claimsIdentity = this.createClaimsIdentity(botAppId);
+        claimsIdentity.claims.push({ type: AuthenticationConstants.ServiceUrlClaim, value: serviceUrl });
```
Comment (stevengum) [inline libraries/botbuilder-core/src/cloudAdapterBase.ts:195]: This implementation is missing the argument validation of the `serviceUrl`, `conversationParameters` and `callback` parameters from [.NET](https://github.com/microsoft/botbuilder-dotnet/blob/001a8a19ba2f4c05ad5149f05689fb6c13495bf4/libraries/Microsoft.Bot.Builder/CloudAdapterBase.cs#L198-L204), is this intentional?

File: libraries/botbuilder-core/src/cloudAdapterBase.ts [hunk 2]
```diff
@@ -176,6 +179,82 @@ export abstract class CloudAdapterBase extends BotAdapter {
...
+            conversationParameters
+        );
+
+        // Create a UserTokenClient instance for the application to use. (For example, in the OAuthPrompt.)
+        const userTokenClient = await this.botFrameworkAuthentication.createUserTokenClient(claimsIdentity);
+
+        // Create a turn context and run the pipeline.
+        const context = this.createTurnContext(
+            createActivity,
+            claimsIdentity,
```
Comment (stevengum) [inline libraries/botbuilder-core/src/cloudAdapterBase.ts:220]: I would expect that at this point, the `claimsIdentity` would have an audienceClaim with the value of `audience`, not the `botAppId`:

https://github.com/microsoft/botbuilder-js/blob/0db823143a83267c60d42a7daef2cb04cc72ee4a/libraries/botbuilder-core/src/cloudAdapterBase.ts#L307-L315

Is there a reason this isn't the case?

What will happen is that the `ClaimsIdentity` is essentially Emulator-esque as the appId and audience claim are the same value (the `botAppId`), whereas the created `ConnectorClient` is properly addressed to the passed in `audience`. This `audience` will can be a channel, another bot, or something else entirely.

File: libraries/botbuilder-core/src/cloudAdapterBase.ts [hunk 3]
```diff
@@ -176,6 +179,82 @@ export abstract class CloudAdapterBase extends BotAdapter {
...
+            logic,
+            connectorFactory
+        );
+
+        // Run the pipeline.
+        await this.runMiddleware(context, logic);
+    }
+
+    private createCreateActivity(
+        createdConversationId: string | undefined,
```
Comment (stevengum) [inline libraries/botbuilder-core/src/cloudAdapterBase.ts:233]: Thanks for only taking in the necessary data, not the entire ConversationResourceResponse!

**Review by stevengum — COMMENTED** [review]
Are the applicable Teams flows functional with these changes?

File: libraries/botbuilder-core/src/cloudAdapterBase.ts (hunk 2 above)
Comment (joshgummersall) [reply libraries/botbuilder-core/src/cloudAdapterBase.ts:220]: Not sure, I followed John's lead in porting this. As far as I can tell, this is equivalent [to .NET](https://github.com/microsoft/botbuilder-dotnet/blob/54c75546821c3e4b90d7b24a3f98da549761e753/libraries/Microsoft.Bot.Builder/CloudAdapterBase.cs#L321). Perhaps we should file an issue on .NET?

File: libraries/botbuilder-core/src/cloudAdapterBase.ts (hunk 1 above)
Comment (joshgummersall) [reply libraries/botbuilder-core/src/cloudAdapterBase.ts:195]: Laziness on my part, will fix!

**Comment by joshgummersall** [comment]
Working on testing in Teams.

**Review by stevengum — APPROVED** [review]
Approved, let's follow up on the proper claims construction in the .NET repo

**Pull request merged**